File: companies/Huawei/29.divide-two-integers.cpp
#include <algorithm>
#include <cctype>
#include <climits>
#include <cmath>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>
using namespace std;


class Solution {
public:
    int divide(int dividend, int divisor) {
        /*
        Approach:
        Work with positive long long magnitudes and subtract the largest shifted
        divisor chunk possible at each step. This is binary long division using
        bit shifts instead of multiplication, division, or modulo.

        C++ notes:
        long long safely holds abs(INT_MIN). The only overflowing 32-bit result
        is INT_MIN / -1, which must clamp to INT_MAX.

        Complexity: O(log |dividend|) time and O(1) space.
        */
        if (dividend == INT_MIN && divisor == -1) return INT_MAX;
        long long a = llabs((long long)dividend);
        long long b = llabs((long long)divisor);
        long long quotient = 0;
        while (a >= b) {
            long long chunk = b, multiple = 1;
            while ((chunk << 1) <= a) {
                chunk <<= 1;
                multiple <<= 1;
            }
            a -= chunk;
            quotient += multiple;
        }
        bool negative = (dividend < 0) ^ (divisor < 0);
        return (int)(negative ? -quotient : quotient);
    }
};

File: companies/Huawei/29.divide-two-integers.go
package leetcode

// Divide29 performs binary long division by repeatedly subtracting the largest
// doubled divisor chunk that fits. It works with absolute values, restores sign,
// and handles the one 32-bit overflow case.
//
// Time: O(log^2 N) in this simple doubling form. Space: O(1).
func Divide29(dividend int, divisor int) int {
	const intMax = 1<<31 - 1
	const intMin = -1 << 31
	if dividend == intMin && divisor == -1 {
		return intMax
	}
	negative := (dividend < 0) != (divisor < 0)
	a, b := dividend, divisor
	if a < 0 {
		a = -a
	}
	if b < 0 {
		b = -b
	}
	q := 0
	for a >= b {
		chunk, multiple := b, 1
		for a >= chunk<<1 {
			chunk <<= 1
			multiple <<= 1
		}
		a -= chunk
		q += multiple
	}
	if negative {
		return -q
	}
	return q
}
